### PR TITLE
editoast: add infragraph

### LINKED ISSUE #1034: Generate a topological graph of the infra
Build an infragraph with a struct like :
```rust
struct TracksectionGraph {
    id : String
    neighbors_start: Vec<String>
    neighbors_stop: Vec<String>
}
```
or use a lib rust ?

### PR DESCRIPTION
**Pull request by Tguisnet**
close #1034

### CODE REVIEW

File: editoast/src/models/errors/track_section_links.rs [hunk 1]
```diff
@@ -17,15 +17,21 @@ pub fn generate_errors(
...
             let infra_error = InfraError::new_invalid_reference("src.track".into(), obj_ref);
             errors.push(to_value(infra_error).unwrap());
             link_ids.push(link_id.clone());
         }
 
-        if !infra_cache.track_sections.contains_key(&link.dst) {
-            let obj_ref = ObjectRef::new(ObjectType::TrackSection, link.dst.clone());
+        if !infra_cache
+            .track_sections
+            .contains_key(&link.dst.track.obj_id)
```
Comment (flomonster) [inline editoast/src/models/errors/track_section_links.rs]: You can factorize it

File: editoast/src/models/errors/graph.rs [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+    fn create_track_endpoint(endpoint: Endpoint, obj_id: String) -> TrackEndpoint {
+        TrackEndpoint {
+            endpoint,
+            track: ObjectRef {
+                obj_type: ObjectType::TrackSection,
+                obj_id,
+            },
+        }
+    }
```
Comment (flomonster) [inline editoast/src/models/errors/graph.rs]: why not implementing a `new` function to `TrackEndpoint`, `TrackSectionLink`, `SwitchPortConnection`...?

File: editoast/src/models/errors/graph.rs [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        );
+        groups.insert(
+            "RIGHT".into(),
+            vec![create_switch_connection("BASE".into(), "RIGHT".into())],
+        );
+        SwitchType {
+            id: "point".into(),
+            ports: vec!["BASE".into(), "LEFT".into(), "RIGHT".into()],
+            groups,
+        }
```
Comment (flomonster) [inline editoast/src/models/errors/graph.rs]: You can define the json using `json!` then parse it.

File: editoast/src/models/errors/graph.rs [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    }
+
+    fn create_switch_cache_point(
+        obj_id: String,
+        base: TrackEndpoint,
+        left: TrackEndpoint,
+        right: TrackEndpoint,
+    ) -> SwitchCache {
+        let mut ports = HashMap::default();
```
Comment (flomonster) [inline editoast/src/models/errors/graph.rs]: You can use `map!` to create and fill the hashmap at the same time

File: editoast/src/models/errors/graph.rs (hunk 2 above)
Comment (flomonster) [inline editoast/src/models/errors/graph.rs]:
```suggestion
    fn create_track_endpoint<T: AsRef<str>>(endpoint: Endpoint, obj_id: T) -> TrackEndpoint {
        TrackEndpoint {
            endpoint,
            track: ObjectRef {
                obj_type: ObjectType::TrackSection,
                obj_id.as_ref().into(),
            },
        }
    }
```

With this implementation you can call this function with `&str` or `String` or `&String`...

File: editoast/src/models/errors/graph.rs [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+            }
+        }
+
+        // second branch
+        let ed = create_track_endpoint(Endpoint::End, "B".into());
+        if let Some(branch) = graph.links.get(&ed) {
+            if let Some(compared_branch) = graph.get_neighbours(&ed) {
+                assert_eq!(compared_branch, branch);
+            }
+        }
```
Comment (flomonster) [inline editoast/src/models/errors/graph.rs]: I don't agree with these tests.

- You should `expect` some branch and no test it. If you can't find a branch in your graph then we have an issue.
- All the endpoints should be tested. You can at least check that the number of neighbors is correct.

File: editoast/src/railjson/mod.rs [hunk 6]
```diff
@@ -323,15 +323,15 @@ pub enum Side {
...
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq, Hash)]
 pub enum Endpoint {
     #[serde(rename = "BEGIN")]
     Begin,
     #[serde(rename = "END")]
     End,
 }
 
-#[derive(Debug, Derivative, Clone, Deserialize, Serialize)]
+#[derive(Debug, Derivative, Clone, Deserialize, PartialEq, Eq, Hash, Serialize)]
```
Comment (flomonster) [inline editoast/src/railjson/mod.rs]:
```suggestion
#[derive(Debug, Derivative, Clone, Deserialize, Serialize, PartialEq, Eq, Hash)]
```

**Review by flomonster — CHANGES_REQUESTED** [review]

File: editoast/src/models/errors/graph.rs (hunk 3 above)
Comment (Tguisnet) [reply editoast/src/models/errors/graph.rs]: Static Hashmap doesn't exist, maybe use lazy_static ? (overkill ?)

File: editoast/src/infra_cache.rs [hunk 7]
```diff
@@ -728,3 +722,125 @@ impl InfraCache {
...
+
+    /*pub const switch_type_point: SwitchType = SwitchType {
+        id: "point".into(),
+        ports: vec!["BASE".into(), "LEFT".into(), "RIGHT".into()],
+        groups: [(
+            "LEFT".into(),
+            vec![create_switch_connection("BASE".into(), "LEFT".into())]), (
+                "RIGHT".into(),
+                vec![create_switch_connection("BASE".into(), "RIGHT".into())])].into_iter().collect()
+    };*/
```
Comment (flomonster) [inline editoast/src/infra_cache.rs]: dead code

File: editoast/src/infra_cache.rs [hunk 8]
```diff
@@ -728,3 +722,125 @@ impl InfraCache {
...
+        );
+        groups.insert(
+            "RIGHT".into(),
+            vec![create_switch_connection("BASE".into(), "RIGHT".into())],
+        );
+        SwitchType {
+            id: "point".into(),
+            ports: vec!["BASE".into(), "LEFT".into(), "RIGHT".into()],
+            groups,
+        }
```
Comment (flomonster) [inline editoast/src/infra_cache.rs]: use `hash!`

File: editoast/src/models/errors/graph.rs [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    /// Add a new link to the graph given two `TrackEndpoint`.
+    fn link(&mut self, src: &'a TrackEndpoint, dst: &'a TrackEndpoint) {
+        self.links.entry(src).or_default().insert(dst);
+    }
+
+    /// Create a graph from an `InfraCache`.
+    pub fn load(infra_cache: &'a InfraCache) -> Self {
+        let mut graph = Self::default();
+
+        for (_link_id, link) in infra_cache.track_section_links.iter() {
```
Comment (flomonster) [inline editoast/src/models/errors/graph.rs]:
```suggestion
        for link in infra_cache.track_section_links.values() {
```

File: editoast/src/models/errors/graph.rs [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+            match link.navigability {
+                crate::railjson::ApplicableDirections::StartToStop => {
```
Comment (flomonster) [inline editoast/src/models/errors/graph.rs]: Import `railjson::ApplicableDirections`

File: editoast/src/models/errors/graph.rs [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+                    graph.link(&link.dst, &link.src);
+                }
+                crate::railjson::ApplicableDirections::Both => {
+                    graph.link(&link.src, &link.dst);
+                    graph.link(&link.dst, &link.src);
+                }
+            }
+        }
+
+        for (_switch_id, switch) in infra_cache.switches.iter() {
```
Comment (flomonster) [inline editoast/src/models/errors/graph.rs]:
```suggestion
        for switch in infra_cache.switches.values() {
```

File: editoast/src/models/errors/graph.rs [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+            if let Some(switch_type) = infra_cache.switch_types.get(&switch.switch_type) {
+                for (_, group) in switch_type.groups.iter() {
+                    for connection in group {
+                        if let Some(src) = switch.ports.get(&connection.src) {
+                            if let Some(dst) = switch.ports.get(&connection.dst) {
```
Comment (flomonster) [inline editoast/src/models/errors/graph.rs]: use `unwrap()`

File: editoast/src/models/errors/graph.rs [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+                }
+                crate::railjson::ApplicableDirections::Both => {
+                    graph.link(&link.src, &link.dst);
+                    graph.link(&link.dst, &link.src);
+                }
+            }
+        }
+
+        for (_switch_id, switch) in infra_cache.switches.iter() {
+            if let Some(switch_type) = infra_cache.switch_types.get(&switch.switch_type) {
```
Comment (flomonster) [inline editoast/src/models/errors/graph.rs]: use `unwrap`

**Review by flomonster — CHANGES_REQUESTED** [review]

File: editoast/src/models/errors/graph.rs [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+        let all_track_endpoints = vec![
+            &track_a_begin,
+            &track_a_end,
+            &track_b_begin,
+            &track_b_end,
+            &track_c_begin,
+            &track_c_end,
+            &track_d_begin,
+            &track_d_end,
+        ];
```
Comment (flomonster) [inline editoast/src/models/errors/graph.rs]: You can use loop instead of having to list all possible endpoints.

**Review by flomonster — DISMISSED** [review]

**Pull request merged**